dali/pipeline/operators/transpose/transpose_test.cc

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "dali/test/dali_operator_test.h"
+#include "dali/test/dali_operator_test_utils.h"
+#include "dali/pipeline/operators/transpose/transpose.h"
+
+namespace dali {
+
+namespace {
+
+// Fill tensors of consecutive numbers
+template <typename T>
+void Arrange(T* ptr, const std::vector<Index>& shape) {
+  auto volume = Volume(shape);
+  for (int i = 0; i < volume; ++i) {
+    ptr[i] = static_cast<int>(i);
+  }
+}
+
+std::vector<int> GetStrides(const std::vector<Index>& shape) {
+  std::vector<int> strides(shape.size(), 1);
+  for (int i = shape.size() - 2; i >= 0; --i) {
+    strides[i] = shape[i + 1] * strides[i + 1];
+  }
+  return strides;
+}
+
+}  // namespace
+
+class TransposeTest : public testing::DaliOperatorTest {
+  testing::GraphDescr GenerateOperatorGraph() const noexcept override {
+    return {"Transpose"};
+  }
+
+ public:
+  std::unique_ptr<TensorList<CPUBackend>> GetInput(int rank) {
+    constexpr int batch_size = 10;
+    std::vector<Index> seq_shape{4, 8, 6};
+    if (rank == 4) {
+      seq_shape.push_back(2);
+    }
+    std::vector<std::vector<Index>> batch_shape;
+    for (int i = 0; i < batch_size; i++) {
+      batch_shape.push_back(seq_shape);
+    }
+
+    std::unique_ptr<TensorList<CPUBackend>> tl(new TensorList<CPUBackend>);
+    tl->Resize(batch_shape);
+    tl->set_type(TypeInfo::Create<int>());
+
+    for (int i = 0; i < batch_size; i++) {
+      Arrange(tl->mutable_tensor<int>(i), seq_shape);
+    }
+    return tl;
+  }
+};
+
+class TransposeTestRank3 : public TransposeTest {};
+class TransposeTestRank4 : public TransposeTest {};
+
+std::vector<testing::Arguments> permutationsRank3 = {
+    {{"perm", std::vector<int>{0, 1, 2}}},
+    {{"perm", std::vector<int>{1, 2, 0}}},
+    {{"perm", std::vector<int>{2, 1, 0}}},
+    {{"perm", std::vector<int>{2, 0, 1}}},
+    {{"perm", std::vector<int>{0, 2, 1}}},
+    {{"perm", std::vector<int>{1, 0, 2}}},
+};
+
+std::vector<testing::Arguments> permutationsRank4 = {
```

Comment:
I would be lazy here: https://en.cppreference.com/w/cpp/algorithm/next_permutation